### PR TITLE
docs: detect and fix broken links

### PR DESCRIPTION
**Pull request by snitin315**
#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

[x] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[ ] Add something to the core
[ ] Other, please explain:

#### What changes did you make? (Give an overview)

Resolve https://github.com/eslint/eslint/issues/16833

- Enable [hyperlink](https://www.npmjs.com/package/hyperlink) to detect broken links via `npm run lint:links`
- Add a validate links check on CI to run this script.
- Skip/ignore linting the following links/errors:
   - All external links `--skip https://`
   - All nav bar links since the code for them is in a different repo. (`--skip /blog --skip /play --skip /team --skip /donate --skip /docs/latest`)
   - All links on old migration guides `--skip migrating-to`.
   - All the fragment-redirect errors since we use [pretty URLs without `.html`](https://github.com/eslint/eslint/blob/13552c4e8d98c3b38b7dbff9c74dc82ca2aad385/docs/.eleventy.js#L134). (`--skip fragment-redirect`)
   
        Screenshot 2023-01-31 at 6 59 39 AM
        
   - [`null` src](https://github.com/eslint/eslint/blob/13552c4e8d98c3b38b7dbff9c74dc82ca2aad385/docs/.eleventy.js#L243) for links in further reading section. For example[ this one ](https://github.com/eslint/eslint/blob/13552c4e8d98c3b38b7dbff9c74dc82ca2aad385/docs/src/_data/further_reading_links.json#L194). (`--skip _site/rules/null`)
     
        Screenshot 2023-01-31 at 6 42 32 AM

  

#### Is there anything you'd like reviewers to focus on?

The CI time seems to be increased significantly with this check. Should we move docs-related checks into a separate job?

### CODE REVIEW

**Comment by bmish** [comment]
There's also [markdown-link-check](https://www.npmjs.com/package/markdown-link-check) which is more popular on npm. Between that and [hyperlink](https://www.npmjs.com/package/hyperlink), which one do you think is better for us?

**Comment by snitin315** [comment]
[markdown-link-check](https://www.npmjs.com/package/markdown-link-check) seems to work with `.md` files only, however not all of our links are in markdown files. For example, the reported broken link in #16833 is in a json file:

https://github.com/eslint/eslint/blob/13552c4e8d98c3b38b7dbff9c74dc82ca2aad385/docs/src/_data/rules.json#L1896

So I believe `markdown-link-check` won't be able to detect such broken links. Whereas `hyperlink` runs on the generated HTML files so it can detect such links.

Screenshot 2023-01-29 at 10 58 16 AM

**Comment by snitin315** [comment]
Seeing a strange [issue on CI,](https://github.com/eslint/eslint/actions/runs/4071386033/jobs/7013138880) (`FATAL ERROR: Reached heap limit Allocation failed - JavaScript heap out of memory`) debugging it.

Screenshot 2023-02-02 at 10 33 54 AM

File: .github/workflows/ci.yml
Comment (snitin315) [inline .github/workflows/ci.yml:43]:
```suggestion
        NODE_OPTIONS: --max-old-space-size=4096
```

File: docs/package.json
Comment (snitin315) [inline docs/package.json:20]:
```suggestion
        "lint:links": "hyperlink ./_site/index.html -c 8 --canonicalroot https://eslint.org/docs/head/ -r --internal ./_site/index.html --pretty --skip https:// --skip fragment-redirect --skip migrating-to --skip /blog --skip /play --skip /team --skip /donate --skip /docs/latest --skip _site/rules/null | tap-spot",
```

File: docs/package.json
Comment (snitin315) [inline docs/package.json:20]:
```suggestion
        "lint:links": "hyperlink ./_site/index.html -c 50 --canonicalroot https://eslint.org/docs/head/ -r --internal ./_site/index.html --pretty --skip https:// --skip fragment-redirect --skip migrating-to --skip /blog --skip /play --skip /team --skip /donate --skip /docs/latest --skip _site/rules/null | tap-spot",
```

**Review by amareshsm — APPROVED** [review]
LGTM  🎉

**Comment by snitin315** [comment]
> Can we do the same for eslint.org as well?

@harish-sethuraman sure, I'll send a PR to the eslint.org repo as well 👍🏻

**Comment by nzakas** [comment]
Folks - please remember, if you approve a PR but don't merge it, explain why you didn't merge it. If you want someone else specific to review it, at-mention them; if you want another approval before merging, just say that. That way, we can know if something is ready to merge. We'd like to get PRs merged as quickly as possible and this will really help. Thanks!

**Comment by snitin315** [comment]
It seems the merging is blocked, and since I can't approve my own PR I would need another approval from @eslint/eslint-tsc to enable merging.

**Review by nzakas — APPROVED** [review]

**Comment by nzakas** [comment]
We just need to wait for after a potential patch release before merging.

**Comment by nzakas** [comment]
I tried running this locally before merging, and it doesn't appear to work:

```
$ npm run lint:links

> docs-eslint@8.33.0 lint:links
> hyperlink ./_site/index.html --canonicalroot https://eslint.org/docs/head/ -r --internal ./_site/index.html --pretty --skip https:// --skip fragment-redirect --skip migrating-to --skip /blog --skip /play --skip /team --skip /donate --skip /docs/latest --skip _site/rules/null > internal-links.tap; cat internal-links.tap | tap-spot

Guessing --root from input files: file:///C:/Users/nzaka/projects/eslint/eslint/docs
  

  0 tests
  0 passed
```

I am running from within `/docs` and verified that `docs/_site` exists. Is there another step required?

**Comment by snitin315** [comment]
That should work. If you are using windows, I'm not sure if the `> internal-links.tap; cat internal-links.tap | tap-spot`  part of the script is compatible with windows. Can you remove that and run the script and run it again?

**Comment by nzakas** [comment]
Yup, now I'm seeing output. Rather than using tap-spot, it would probably make sense to create a JS file in the `tools` directory that runs Hyperlink, collects the output, and displays it in a pretty format. That way it'll work across all operating systems.

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks! I'm leaving this open for others to review again because it's significantly different now.

**Comment by mdjermanovic** [comment]
Also, there are merge conflicts in `docs/package.json`.

**Comment by snitin315** [comment]
@nzakas Can you once confirm if you are able to run it locally now?

File: docs/src/_data/rules.json
Comment (mdjermanovic) [inline docs/src/_data/rules.json:1896]: This file is autogenerated during releases, so we should also update the template:

https://github.com/eslint/eslint/blob/main/conf/rule-type-list.json

Otherwise, next release will overwrite these changes. You can run `node Makefile.js generateRuleIndexPage` to verify.

**Review by mdjermanovic — CHANGES_REQUESTED** [review]

File: docs/src/_data/rules.json
Comment (snitin315) [reply docs/src/_data/rules.json:1896]: Thanks for the catch, fixed in https://github.com/eslint/eslint/pull/16837/commits/946173f909348caf95760a01d8d52011adffb1d7

**Review by mdjermanovic — APPROVED** [review]
LGTM, thanks! Leaving this open for @nzakas to verify.

**Review by nzakas — APPROVED** [review]
Confirmed it works for me. Thanks!

**Pull request merged**
